ocaml/xapi/taskHelper.ml

```diff
@@ -178,36 +189,43 @@ let status_is_completed task_status =
   task_status = `success || task_status = `failure || task_status = `cancelled
 
 let complete ~__context result =
-  Context.complete_tracing __context ;
-  operate_on_db_task ~__context (fun self ->
-      let status = Db_actions.DB_Action.Task.get_status ~__context ~self in
-      if status = `pending then (
-        Db_actions.DB_Action.Task.set_allowed_operations ~__context ~self
-          ~value:[] ;
-        Db_actions.DB_Action.Task.set_finished ~__context ~self
-          ~value:(Date.of_float (Unix.time ())) ;
-        Db_actions.DB_Action.Task.set_progress ~__context ~self ~value:1. ;
-        set_result_on_task ~__context self result ;
-        Db_actions.DB_Action.Task.set_status ~__context ~self ~value:`success
-      ) else
-        debug "the status of %s is: %s; cannot set it to `success"
-          (Ref.really_pretty_and_small self)
-          (status_to_string status)
-  )
+  Xapi_stdext_pervasives.Pervasiveext.finally
+    (fun () ->
+      operate_on_db_task ~__context (fun self ->
+          let status = Db_actions.DB_Action.Task.get_status ~__context ~self in
+          if status = `pending then (
+            Db_actions.DB_Action.Task.set_allowed_operations ~__context ~self
+              ~value:[] ;
+            Db_actions.DB_Action.Task.set_finished ~__context ~self
+              ~value:(Date.of_float (Unix.time ())) ;
```

Comment:
We are changing time handling; @psafont might want to take a look here.

Reply:
Since it's in the database, and it's synchronized between hosts, it should be a date, but Unix shouldn't be used. Better to use `(Date.now ())`

Reply:
It's not something related with this PR. I can make a separate one to address this.

Reply:
Done https://github.com/xapi-project/xen-api/pull/5735/commits/2f20af93e7ab6d5683c2e2e27d6dd8be9c2365d2!

Reply:
Ptime and Unix are equivalent in this regard (in terms of retrieving the timestamp, not in terms of converting it to string), but unfortunately we don't store the full precision Ptime.t. See my other comment. Something to be improved in another PR.